sdk/program/src/vote/state/vote_state_deserialize.rs

```diff
@@ -119,15 +133,42 @@ fn read_prior_voters_into<T: AsRef<[u8]>>(
 
 fn read_epoch_credits<T: AsRef<[u8]>>(
     cursor: &mut Cursor<T>,
-) -> Result<Vec<(Epoch, u64, u64)>, InstructionError> {
+) -> Result<Vec<EpochCreditsItem>, InstructionError> {
     let epoch_credit_count = read_u64(cursor)? as usize;
-    let mut epoch_credits = Vec::with_capacity(epoch_credit_count.min(MAX_EPOCH_CREDITS_HISTORY));
+    if epoch_credit_count > MAX_EPOCH_CREDITS_HISTORY {
+        return Err(InstructionError::InvalidAccountData);
```

Comment:
Can we bail out here? I think the bincode-derive generated
::deserialize will happily read more than MAX_EPOCH_CREDITS_HISTORY given such
an (handcrafted) account?

We _can_ bail out if we think that all ::deserialize() calls are preceded by
VoteStateXXX::is_correct_size_and_initialized but is that the case?

Reply:
I'm not following the point you're making here. Why wouldn't we want to bail early if the epoch credits count value was invalid? I notice we have the same `min(MAX..)` logic inside read votes as well.

Reply:
`min()` just clamps the allocation, it doesn't return early. If we return early we add an error condition that isn't present with normal deserialize.

Reply:
Oh I see, you're pointing out that this behavior diverges from normal deserialize. Ok, I'll revert back to the normal behavior.

Reply:
Ok well I have a dilemma here because I don't want to allocate the vector with more than the max capacity because if someone deserializes untrusted bytes they could get allocate a ton of memory here. I think I have to keep the code as-is.. which should be fine because we only use it in the runtime on vote program owned accounts which cannot have `epoch_credit_count` exceed `MAX_EPOCH_CREDITS_HISTORY`.

Reply:
> Ok well I have a dilemma here because I don't want to allocate the vector with more than the max capacity because if someone deserializes untrusted bytes they could get allocate a ton of memory here.

Yeah, I think if you wanted to not error out, you could fallback on not doing the optimization but...

> which should be fine because we only use it in the runtime on vote program owned accounts which cannot have epoch_credit_count exceed MAX_EPOCH_CREDITS_HISTORY

Agreed, you've talked me into this before and I'm convinced 😅

Reply:
> Yeah, I think if you wanted to not error out, you could fallback on not doing the optimization but...

Oh yeah, duh. I'll just do that, thanks!